### PR TITLE
[image_picker] add requestFullMetadata for iOS (optional permissions)

### PR DESCRIPTION
**Pull request by PiotrMitkowski**
## Description
Originally #3264 made by @cpboyd 

Make iOS 11+ permissions requests optional per https://developer.apple.com/forums/thread/653414.
PHAssets aren't actually required and there was already some fallback code in place.

## Related issues
Fixes flutter/flutter#65995

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/image_picker/image_picker/lib/image_picker.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker/lib/image_picker.dart:194]: Nit: remove the comma after "control".

Comment (stuartmorgan) [inline packages/image_picker/image_picker/lib/image_picker.dart:194]: Nit: This shouldn't specifically say it's for iOS, as it might apply to other platforms in the future. I would remove iOS here, and add it to the sentence below (e.g., "... which may require extra permission requests on some platforms, such as iOS").

File: packages/image_picker/image_picker_platform_interface/CHANGELOG.md
Comment (stuartmorgan) [inline packages/image_picker/image_picker_platform_interface/CHANGELOG.md:3]: Re-adds

(Per style guide)

File: packages/image_picker/image_picker_platform_interface/CHANGELOG.md
Comment (stuartmorgan) [inline packages/image_picker/image_picker_platform_interface/CHANGELOG.md:4]: You don't need to explain in the changelog that it's non-breaking; that's indicated by the version.

File: packages/image_picker/image_picker_platform_interface/lib/src/method_channel/method_channel_image_picker.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_platform_interface/lib/src/method_channel/method_channel_image_picker.dart:277]: Please put this next to `getImage`

File: packages/image_picker/image_picker_platform_interface/lib/src/platform_interface/image_picker_platform.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_platform_interface/lib/src/platform_interface/image_picker_platform.dart:149]: Nit: Missing final period.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Where are the iOS changes? This PR adds all the plumbing, but has no effect on any platform.

File: packages/image_picker/image_picker_platform_interface/CHANGELOG.md
Comment (PiotrMitkowski) [reply packages/image_picker/image_picker_platform_interface/CHANGELOG.md:4]: Okay, I only wanted to explicitly show the difference from previous implementation

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m
Comment (stuartmorgan) [inline packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:100]: The type is `PHAsset`, not `PHAAsset`, and also ObjC style is to capitalize abbreviations even in variable names, so this, and the similar variables below, should be `usePHAsset` rather than `usePhaAsset`.

File: packages/image_picker/image_picker/lib/image_picker.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker/lib/image_picker.dart:194]: Remove the "on iOS" here.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
The iOS changes need unit tests that the parameter is actually honored (since this flow can't use integration tests). Right now all of the iOS code here could be broken or removed, and no tests would fail, which means there's no meaningful testing.

**Comment by PiotrMitkowski** [comment]
@stuartmorgan can you give me a little hint on the iOS tests? I see two expected results of having the `requestFullMetadata` flag set to `false`:

1. when running `pickImage` method call, plugin's method  `checkPhotoAuthorization` should not be called
2. after picking an image, plugin's method `saveImageWithPickerInfo` should be called

For now I don't see a way to check if any of those happened in the iOS unit tests.

**Comment by stuartmorgan** [comment]
> I see two expected results of having the `requestFullMetadata` flag set to `false`:

Those aren't expected results, they are internal implementation details. Expected results here would be expressed in terms of the iOS APIs that are or are not called, which would be checked by mocking them.

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m
Comment (stuartmorgan) [inline packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:159]: According to the thread you linked to, this change is only correct on iOS 11+, and iOS 9 and 10 still need permissions.

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m
Comment (stuartmorgan) [inline packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m:186]: I'm confused by this change. Why is forcing `UIImagePicker` desirable, given that `PHPicker` also shouldn't require permission prompting? Isn't the permission issue specifically related to accessing metadata, not the top-level API used?

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m
Comment (PiotrMitkowski) [reply packages/image_picker/image_picker/ios/Classes/FLTImagePickerPlugin.m]: You're right, for `PHPicker` it's enough to prevent from calling util method `getAssetFromImagePickerInfo`. I'll fix that part

**Comment by cedvdb** [comment]
is the default to not request full metadata ? It should

File: packages/image_picker/image_picker/lib/image_picker.dart
Comment (cedvdb) [inline packages/image_picker/image_picker/lib/image_picker.dart:213]: The default should be false, it makes more sense even if it is a breaking change for some.

Comment (stuartmorgan) [reply packages/image_picker/image_picker/lib/image_picker.dart:213]: > Imo, the default should be false, it makes more sense

If this were a new API, I would agree it should default to false, but it's not. We wouldn't make a breaking change to the plugin just for this.

Feel free to file an issue to track changing the default the next time we decide to make a breaking change for other reasons.

> even if it is a breaking change for some.

There's no such thing as a breaking change "for some" in a plugin. A breaking change will change the major version, which affects every client of the plugin whether they would be broken by the specific change or not.

File: packages/image_picker/image_picker_platform_interface/lib/src/types/image_picker_options.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_platform_interface/lib/src/types/image_picker_options.dart:38]: The wrapping on this comment is very uneven; please wrap everything to 80 chars.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
It looks like you've broken one of the existing iOS tests:
```
Failing tests:
	RunnerUITests:
		-[ImagePickerFromGalleryUITests testPickingFromGallery]
```

File: packages/image_picker/image_picker_android/pubspec.yaml
Comment (cyanglaz) [inline packages/image_picker/image_picker_android/pubspec.yaml:34]: add a line to the end of file

File: packages/image_picker/image_picker_ios/example/ios/RunnerTests/ImagePickerPluginTests.m
Comment (cyanglaz) [inline packages/image_picker/image_picker_ios/example/ios/RunnerTests/ImagePickerPluginTests.m:193]: nits: @(NO)

File: packages/image_picker/image_picker_ios/example/pubspec.yaml
Comment (cyanglaz) [inline packages/image_picker/image_picker_ios/example/pubspec.yaml:35]: line EOF

File: packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m
Comment (cyanglaz) [inline packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m:136]: Should the BOOL name to be `requestFullMetadata`? Or explain why `usePHAsset` is the same as `requestFullMetadata` in a comment

File: packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m
Comment (cyanglaz) [inline packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m:150]: ditto

File: packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m
Comment (cyanglaz) [inline packages/image_picker/image_picker_ios/ios/Classes/FLTImagePickerPlugin.m:223]: ditto

**Review by cyanglaz — CHANGES_REQUESTED** [review]

File: packages/image_picker/image_picker_android/pubspec.yaml
Comment (stuartmorgan) [reply packages/image_picker/image_picker_android/pubspec.yaml:34]: @cyanglaz Path-based dependency overrides are always temporary lines; this whole block will be reverted before anything can land, in step 4 of [the instructions](https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changing-federated-plugins). Given that, we shouldn't worry about whitespace on those bits.

File: packages/image_picker/image_picker_platform_interface/lib/src/platform_interface/image_picker_platform.dart
Comment (cyanglaz) [inline packages/image_picker/image_picker_platform_interface/lib/src/platform_interface/image_picker_platform.dart:274]: ditto: can we make `options` non-null?

**Review by cyanglaz — CHANGES_REQUESTED** [review]

File: packages/image_picker/image_picker_platform_interface/lib/src/platform_interface/image_picker_platform.dart
Comment (PiotrMitkowski) [reply packages/image_picker/image_picker_platform_interface/lib/src/platform_interface/image_picker_platform.dart:273]: This parameter is needed whenever the picking method is called. All parameters inside `ImagePickerOptions ` are more optional, so the whole options object can be skipped if a caller doesn't need any specific adjustments. It was @stuartmorgan's suggestion in this PR that I've agreed with, as calling the picking method doesn't require creating the options object for every use

**Review by cyanglaz — COMMENTED** [review]
LGTM, @stuartmorgan Do you mind taking another look before we update the path dependency?

File: packages/image_picker/image_picker_platform_interface/lib/src/types/image_picker_options.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_platform_interface/lib/src/types/image_picker_options.dart:19]: You don't need to include the name of a field in the field comment name (as you do in the context of a method comment, where this was taken from). Also, the second sentence isn't true once you split it apart into two different comments for the two fields, because if one maximum is specified, it'll be scaled to fit that max which changes the other dimension. So it should be:

> The maximum width of the image, in pixels. If `null`, the image will only be resized if `maxHeight` is specified.

File: packages/image_picker/image_picker_platform_interface/lib/src/types/image_picker_options.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_platform_interface/lib/src/types/image_picker_options.dart:24]: Same here.

File: packages/image_picker/image_picker_platform_interface/lib/src/types/image_picker_options.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_platform_interface/lib/src/types/image_picker_options.dart:28]: "If null, the image will be returned at its original quality."

Comment (stuartmorgan) [reply packages/image_picker/image_picker_platform_interface/lib/src/types/image_picker_options.dart:28]: Also, lets put a blank comment line after the first sentence here, per https://dart.dev/guides/language/effective-dart/documentation#do-separate-the-first-sentence-of-a-doc-comment-into-its-own-paragraph

File: packages/image_picker/image_picker_platform_interface/lib/src/types/image_picker_options.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_platform_interface/lib/src/types/image_picker_options.dart:10]: Remove "Any of the params may be null"; nullability is part of types now, and doesn't need to be commented.

File: packages/image_picker/image_picker_platform_interface/lib/src/types/image_picker_options.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_platform_interface/lib/src/types/image_picker_options.dart:38]: We try to avoid commenting about specific platforms in the platform interface (other than as illustrative examples). You can just remove this whole sentence since it's a minor detail that callers at the platform interface level aren't going to need to know.

File: packages/image_picker/image_picker_platform_interface/lib/src/types/image_picker_options.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_platform_interface/lib/src/types/image_picker_options.dart:35]: Remove this sentence, and add a blank line.

File: packages/image_picker/image_picker_platform_interface/lib/src/types/image_picker_options.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_platform_interface/lib/src/types/image_picker_options.dart:36]: "Ignored if the source is not [ImageSource.camera], or the chosen camera is not supported on the device."

File: packages/image_picker/image_picker_platform_interface/lib/src/types/image_picker_options.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_platform_interface/lib/src/types/image_picker_options.dart:42]: A field comment should lead with what it's for, not what the default is. How about:

> If true, requests full image metadata, which may require extra permissions on some platforms, (e.g., `NSPhotoLibraryUsageDescription` on iOS).
>
> Defaults to true.

File: packages/image_picker/image_picker_ios/pigeons/messages.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker_ios/pigeons/messages.dart:38]: Nit: this param would be more idiomatic in ObjC as just `fullMetadata:`; the "with" is implied to go with following arguments, and "with full metadata" reads much better than "with request full metadata".

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
I have some documentation feedback and minor nits, but otherwise this looks good. @PiotrMitkowski please go ahead and open the platform interface PR (with the platform interface parts of my comments below addressed) and we can get that a quick final review and land it to start unwinding the full PR.

File: packages/image_picker/image_picker/lib/image_picker.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker/lib/image_picker.dart:215]: You need to require the version of the platform interface that has this method in pubspec.yaml.

File: packages/image_picker/image_picker/lib/image_picker.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker/lib/image_picker.dart:222]: It would also be better to pull the iOS changes into a separate PR and land it first so that you can require that version, otherwise someone who only updates this package will not actually get a working implementation of the new parameter.

File: packages/image_picker/image_picker/test/image_picker_deprecated_test.dart
Comment (stuartmorgan) [inline packages/image_picker/image_picker/test/image_picker_deprecated_test.dart:61]: You'll need to reconcile the test changes against https://github.com/flutter/plugins/pull/5706 once that lands (which is needed to resolve the current tree breakage, so will be before this PR can land).

File: packages/image_picker/image_picker_ios/pubspec.yaml
Comment (stuartmorgan) [inline packages/image_picker/image_picker_ios/pubspec.yaml:6]: This also needs to require the right version of the platform interface.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

**Comment by PiotrMitkowski** [comment]
I closed this PR by a mistake today, so I've created a new one: https://github.com/flutter/plugins/pull/5915

**Pull request closed without merging**
